elasticgraph-graphql/lib/elastic_graph/graphql/filtering/filter_interpreter.rb

```diff
@@ -195,12 +195,16 @@ def process_any_of_expression(bool_node, expressions, field_path)
             end
           end
 
-          # When our `shoulds` array is empty, the filtering semantics we want is to match no documents.
-          # However, that's not the behavior the datastore will give us if we have an empty array in the
-          # query under `should`. To get the behavior we want, we need to pass the datastore some filter
-          # criteria that will evaluate to false for every document.
-          bool_query = shoulds.empty? ? BooleanQuery::ALWAYS_FALSE_FILTER : BooleanQuery.should(*shoulds)
-          bool_query.merge_into(bool_node)
+          # We want to match the following spemantics for `any_of`:
+          # * filter: {anyOf: []} -> return no results
+          # * filter: {anyOf: [{field: null}]} -> return all results
+          #
+          # Hence, when our `expressions` array is empty, we want to match no documents. However, that's
+          # not the behavior the datastore will give us if we have an empty array in the query under
+          # `should`. To get the behavior we want, we need to pass the datastore some filter criteria
+          # that will evaluate to false for every document.
+          bool_query = expressions.empty? ? BooleanQuery::ALWAYS_FALSE_FILTER : BooleanQuery.should(*shoulds)
+          bool_query.merge_into(bool_node) if expressions.empty? || !shoulds.empty?
         end
```

Comment:
💯 I love how simple this is! I assumed this would be a lot more complicated than it was...

Anyhow, I've got a few suggestions:

```ruby
        # We want to provide the following semantics for `any_of`:
        #
        # * `filter: {anyOf: []}` -> return no results
        # * `filter: {anyOf: [{field: null}]}` -> return all results
        def process_any_of_expression(bool_node, expressions, field_path)
          if expressions.empty?
            # When our `expressions` array is empty, we want to match no documents. However, that's
            # not the behavior the datastore will give us if we have an empty array in the query under
            # `should`. To get the behavior we want, we need to pass the datastore some filter criteria
            # that will evaluate to false for every document.
            BooleanQuery::ALWAYS_FALSE_FILTER.merge_into(bool_node)
            return
          end

          shoulds = expressions.filter_map do |expression|
            build_bool_hash do |inner_bool_node|
              process_filter_hash(inner_bool_node, expression, field_path)
            end
          end

          return if shoulds.empty?

          BooleanQuery.should(*shoulds).merge_into(bool_node)
        end
```

* Instead of checking `expressions.empty?` twice as your implementation does, we can check it once and just exit early without even executing the `shoulds =` logic.
* The comments about `expressions` being empty were previously above a couple of lines that handle multiple cases (through a ternary and a trailing `if`). It's easier to follow if we put those docs just under the `expressions.empty?` case.
* Here I've updated a few things in your doc:
  * `match` -> `provide` as I think that reads better
  * `spemantics` -> `semantics`
  * using backticks around your code snippets

Reply:
Done, implemented the suggestions